source/arch/encoding_simd_sse41.c

```diff
@@ -0,0 +1,396 @@
+#include <ctype.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <aws/common/common.h>
+
+/***** Decode logic *****/
+
+/*
+ * Decodes ranges of bytes in place
+ * For each byte of 'in' that is between lo and hi (inclusive), adds offset and _adds_ it to the corresponding offset in
+ * out.
+ */
+static inline __m256i translate_range(__m256i in, uint8_t lo, uint8_t hi, uint8_t offset) {
+    __m256i lovec = _mm256_set1_epi8(lo);
+    __m256i hivec = _mm256_set1_epi8(hi - lo);
+    __m256i offsetvec = _mm256_set1_epi8(offset);
+
+    __m256i tmp = _mm256_sub_epi8(in, lovec);
+    /*
+     * we'll use the unsigned min operator to do our comparison. Note that
+     * there's no unsigned compare as a comparison intrinsic.
+     */
+    __m256i mask = _mm256_min_epu8(tmp, hivec);
+    /* if mask = tmp, then keep that byte */
+    mask = _mm256_cmpeq_epi8(mask, tmp);
+
+    tmp = _mm256_add_epi8(tmp, offsetvec);
+    tmp &= mask;
+    return tmp;
+}
+
+/*
+ * For each 8-bit element in in, if the element equals match, add to the corresponding element in out the value decode.
+ */
+static inline __m256i translate_exact(__m256i in, uint8_t match, uint8_t decode) {
+    __m256i mask = _mm256_cmpeq_epi8(in, _mm256_set1_epi8(match));
+    return mask & _mm256_set1_epi8(decode);
+}
+
+/*
+ * Input: a pointer to a 256-bit vector of base64 characters
+ * The pointed-to-vector is replaced by a 256-bit vector of 6-bit decoded parts;
+ * on decode failure, returns 0, else returns 1 on success.
+ */
+static inline bool decode_vec(__m256i *in) {
+    __m256i tmp;
+
+    tmp = translate_range(*in, 'A', 'Z', 0 + 1);
+    tmp |= translate_range(*in, 'a', 'z', 26 + 1);
+    tmp |= translate_range(*in, '0', '9', 52 + 1);
+    tmp |= translate_exact(*in, '+', 62 + 1);
+    tmp |= translate_exact(*in, '/', 63 + 1);
+
+    /*
+     * We use 0 to mark decode failures, so everything is decoded to one higher
+     * than normal. We'll shift this down now.
+     */
+    *in = _mm256_sub_epi8(tmp, _mm256_set1_epi8(1));
+
+    /* If any byte is now zero, we had a decode failure */
+    __m256i mask = _mm256_cmpeq_epi8(tmp, _mm256_set1_epi8(0));
+    return _mm256_testz_si256(mask, mask);
+}
+
+struct aligned256 {
+    unsigned char bytes[32];
+} __attribute__((aligned(32)));
+
+/*
+ * Input: a 256-bit vector, interpreted as 32 * 6-bit values
+ * Output: a 256-bit vector, the lower 24 bytes of which contain the packed version of the input
+ */
+static inline __m256i pack_vec(__m256i in) {
+    /*
+     * Our basic strategy is to split the input vector into three vectors, for each 6-bit component
+     * of each 24-bit group, shift the groups into place, then OR the vectors together. Conveniently,
+     * we can do this on a dword-by-dword basis.
+     *
+     * It's important to note that we're interpreting the vector as being little-endian. That is,
+     * on entry, we have 32-bit dwords that looks like this:
```

Comment:
"32-bit dwords" is either wrong (on architectures where bytes are not 8 bits wide, but presumably we don't target these), or redundant (a dword is 4 bytes).

Reply:
We're already in intel-specific-code-land, and this is mostly a reminder to the reader that dword is defined as 32-bits on x86 (it may not be obvious for people who don't do a lot of windows or asm coding).

Reply:
Ok, but then please do so upon the first use of "dword." Don't educate people five lines later.